File: packages/core/src/bindings.ts
import { BindingInventorySchema, type BindingInventory } from "./types/goldenBinding.js";

/** Parse stored bindings JSON, degrading corrupt/schema-drifted input to null (v1.5 contract). */
export function parseBindings(raw: string | null): BindingInventory | null {
  if (!raw) return null;
  try {
    return BindingInventorySchema.parse(JSON.parse(raw)) as BindingInventory;
  } catch {
    return null;
  }
}

/** The single boundness counting rule (engine boundnessLine and the UI both format from this). */
export function boundnessCounts(inv: BindingInventory | null): { bound: number; mismatch: number; total: number } | null {
  if (!inv) return null;
  const bound = inv.items.filter((i) => i.binding?.status === "bound").length;
  const mismatch = inv.items.filter((i) => i.binding?.status === "mismatch").length;
  return { bound, mismatch, total: inv.items.length };
}

File: backend/runoff_api/core/bindings.py
"""Port of packages/core/src/bindings.ts — stored-bindings parse + boundness rule."""

import json

from pydantic import ValidationError

from .types.golden_binding import BindingInventory


def parse_bindings(raw: str | None) -> dict | None:
    """Corrupt/schema-drifted stored bindings degrade to None (v1.5 contract)."""
    if not raw:
        return None
    try:
        return BindingInventory.model_validate(json.loads(raw)).model_dump(by_alias=True, exclude_unset=True)
    except (ValidationError, ValueError):
        return None


def boundness_counts(inv: dict | None) -> dict | None:
    if inv is None:
        return None
    items = inv["items"]
    bound = sum(1 for i in items if (i.get("binding") or {}).get("status") == "bound")
    mismatch = sum(1 for i in items if (i.get("binding") or {}).get("status") == "mismatch")
    return {"bound": bound, "mismatch": mismatch, "total": len(items)}
